packages/komencikit/src/kit.ts

```diff
@@ -0,0 +1,378 @@
+   */
+  @retry({
+    tries: 3,
+    bailOnErrorTypes: [
+      FetchErrorTypes.Unauthorised,
+      FetchErrorTypes.ServiceUnavailable,
+      TxErrorTypes.Revert,
+    ],
+    onRetry: (_args, error, attempt) => {
+      console.debug(`${TAG}/submitMetaTransaction attempt#${attempt} error: `, error)
+    },
+  })
+  public async submitMetaTransaction(
+    walletAddress: string,
+    tx: CeloTransactionObject<any>,
+    nonce?: number
+  ): Promise<Result<TransactionReceipt, FetchError | TxError>> {
+    const wallet = await this.getWallet(walletAddress)
+    const signature = await wallet.signMetaTransaction(tx.txo, nonce)
+    const rawMetaTx = toRawTransaction(wallet.executeMetaTransaction(tx.txo, signature).txo)
+
+    const resp = await this.client.exec(submitMetaTransaction(rawMetaTx))
+    if (resp.ok) {
+      const txHash = resp.result.txHash
+      return this.waitForReceipt(txHash)
+    }
+
+    return resp
+  }
+
+  /**
+   * Utility function used to wait for a transaction to finalise and return the receipt
+   *
+   * @param txHash - the hash of the transaction to watch
+   * @private
+   */
+  private async waitForReceipt(txHash: string): Promise<Result<TransactionReceipt, TxError>> {
+    let receipt: TransactionReceipt | null = null
+    let waited = 0
+    while (receipt == null && waited < this.options.txRetryTimeoutMs) {
+      receipt = await this.contractKit.web3.eth.getTransactionReceipt(txHash)
+      if (receipt == null) {
+        await sleep(100)
+        waited += 100
+      }
+    }
+
+    if (receipt == null) {
+      return Err(new TxTimeoutError())
+    }
+
+    return Ok(receipt)
+  }
+
+  /**
+   * Utility function used to instantiate a MetaTransactionWalletWrapper cached by address
+   *
+   * @param txHash - the hash of the transaction to watch
+   * @private
+   */
+  _wallet?: MetaTransactionWalletWrapper
+  private async getWallet(address: string): Promise<MetaTransactionWalletWrapper> {
+    if (this._wallet === undefined || this._wallet.address !== address) {
+      this._wallet = await this.contractKit.contracts.getMetaTransactionWallet(address)
+    }
+    return this._wallet
+  }
+
+  /**
+   * Used to create a signature for a login message that allows the server to verify that the
+   * externalAccount passed in is actually owned by the caller.
+   *
+   * TODO: In order to increase security and prevent replay attacks we should introduce a nonce
+   * Flow would be:
+   * 1. KomenciKit asks for a nonce from Komenci
+   * 2. Komenci saves the nonce in state, associated with the external account requesting it
+   * 3. KomenciKit signs a message containing the nonce and uses it to call startSession
+   *
+   * @param externalAccount
+   * @returns the signature of the login message
+   * @private
+   */
+  private async getLoginSignature(): Promise<Result<string, LoginSignatureError>> {
+    try {
+      const signature = await this.contractKit.web3.eth.sign(
```

Comment:
We should use EIP712 (signTypedData format) for these types of signatures.

Reply:
If it's significant work, we can roll it in later.

Reply:
Yeah I felt like the data-structure would be to thin to warrant an EIP712. Is there a more general convention about when to use personal message vs EIP712 or is it more like personal message is out we should always use 712?